Content.Shared/Chemistry/SolutionContainerSystem.cs

```diff
@@ -0,0 +1,516 @@
+            {
+                return true;
+            }
+
+            solution = null;
+            return false;
+        }
+
+        public bool TryGetDefaultSolution(IEntity? target,
+            [NotNullWhen(true)] out SolutionAlias? solution)
+        {
+            if (target == null
+                || target.Deleted || !target.TryGetComponent(out SolutionContainerManager? solutionsMgr)
+                || solutionsMgr.Solutions.Count != 1)
+            {
+                solution = null;
+                return false;
+            }
+
+            solution = solutionsMgr.Solutions.Values.ToArray()[0];
+            return true;
+        }
+
+        public SolutionAlias? EnsureSolution(IEntity owner, string name)
+        {
+            if (owner.Deleted || !owner.TryGetComponent(out SolutionContainerManager? solutionsMgr))
+            {
+                Logger.Warning($@"Entity (id:{owner.Uid}) is deleted or has no container manager");
+                return null;
+            }
+
+            if (!solutionsMgr.Solutions.ContainsKey(name))
+            {
+                var newSolution = new SolutionAlias();
+                newSolution.Owner = owner;
+                solutionsMgr.Solutions.Add(name, newSolution);
+            }
+            return solutionsMgr.Solutions[name];
+        }
+
+
+        public bool HasSolution(IEntity owner)
+        {
+            return !owner.Deleted && owner.HasComponent<SolutionContainerManager>();
+        }
+
+        public void AddDrainable(IEntity owner, string name)
+        {
+            if (owner.Deleted)
+                return;
+
+            if (!owner.TryGetComponent(out DrainableSolutionComponent? component))
+            {
+                component = owner.AddComponent<DrainableSolutionComponent>();
+            }
+
+            component.Solution = name;
+        }
+        public void RemoveDrainable(IEntity owner)
+        {
+            if (owner.Deleted || !owner.HasComponent<DrainableSolutionComponent>())
+                return;
+
+            owner.RemoveComponent<DrainableSolutionComponent>();
+        }
+
+        public void AddRefillable(IEntity owner, string name)
+        {
+            if (owner.Deleted)
+                return;
+
+            if (!owner.TryGetComponent(out RefillableSolutionComponent? component))
+            {
+                component = owner.AddComponent<RefillableSolutionComponent>();
+            }
+
+            component.Solution = name;
+        }
+
+        public void RemoveRefillable(IEntity owner)
+        {
+            if (owner.Deleted || !owner.HasComponent<RefillableSolutionComponent>())
+                return;
+
+            owner.RemoveComponent<RefillableSolutionComponent>();
+        }
+
+
+        public IList<string> RemoveEachReagent(SolutionAlias solution, ReagentUnit quantity)
+        {
+            var removedReagent = new List<string>(solution.Contents.Count);
+            if (quantity <= 0)
```

Comment:
Would be better to make an array, that way you can also use Array.Empty<string> if quantity <= 0

Also @Zumorica because it's around plantholder.